File: testers/sqlite/java/src/java/testdb/MockRepoTest.java
package testdb;

import static org.junit.Assert.*;

import dev.typr.dsl.MockConnection;
import dev.typr.foundations.Bijection;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Optional;
import org.junit.Test;
import testdb.customers.*;
import testdb.departments.*;
import testdb.userdefined.Email;

/** Generated mock repositories — exercised entirely in-process, no SQLite connection. */
public class MockRepoTest {
  private static CustomersRepoMock createCustomersMock() {
    return new CustomersRepoMock(
        unsaved -> unsaved.toRow(() -> LocalDateTime.of(2025, 1, 1, 0, 0)));
  }

  private final CustomersRepoMock customersMock = createCustomersMock();
  private final DepartmentsRepoMock departmentsMock = new DepartmentsRepoMock();

  @Test
  public void testMockInsertAndSelect() {
    var customer =
        new CustomersRow(
            new CustomersId(1L),
            "Mock User",
            Optional.of(new Email("mock@test.com")),
            LocalDateTime.of(2025, 1, 1, 0, 0));

    var inserted = customersMock.insert(customer, null);
    assertNotNull(inserted);
    assertEquals("Mock User", inserted.name());

    var found = customersMock.selectById(new CustomersId(1L), null);
    assertTrue(found.isPresent());
    assertEquals("Mock User", found.get().name());
  }

  @Test
  public void testMockUpdate() {
    var customer =
        new CustomersRow(
            new CustomersId(2L),
            "Original Name",
            Optional.empty(),
            LocalDateTime.of(2025, 1, 1, 0, 0));
    customersMock.insert(customer, null);

    var updated = customer.withName("Updated Name");
    customersMock.update(updated, null);

    var found = customersMock.selectById(new CustomersId(2L), null).orElseThrow();
    assertEquals("Updated Name", found.name());
  }

  @Test
  public void testMockDelete() {
    var customer =
        new CustomersRow(
            new CustomersId(3L), "To Delete", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0));
    customersMock.insert(customer, null);
    assertTrue(customersMock.selectById(new CustomersId(3L), null).isPresent());

    customersMock.deleteById(new CustomersId(3L), null);
    assertFalse(customersMock.selectById(new CustomersId(3L), null).isPresent());
  }

  @Test
  public void testMockSelectAll() {
    var mock = createCustomersMock();
    mock.insert(
        new CustomersRow(
            new CustomersId(10L), "U1", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);
    mock.insert(
        new CustomersRow(
            new CustomersId(11L), "U2", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);
    mock.insert(
        new CustomersRow(
            new CustomersId(12L), "U3", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);

    var all = mock.selectAll(null);
    assertEquals(3, all.size());
  }

  @Test
  public void testMockWithCompositeKey() {
    var dept =
        new DepartmentsRow(
            "MOCK_IT", "MOCK_US", "Mock IT US", Optional.of(new BigDecimal("500000")));

    departmentsMock.insert(dept, null);

    var id = new DepartmentsId("MOCK_IT", "MOCK_US");
    var found = departmentsMock.selectById(id, null);

    assertTrue(found.isPresent());
    assertEquals("Mock IT US", found.get().deptName());
  }

  @Test
  public void testMockDSLQuery() {
    var mock = createCustomersMock();

    mock.insert(
        new CustomersRow(
            new CustomersId(100L), "Alice", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);
    mock.insert(
        new CustomersRow(
            new CustomersId(101L), "Bob", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);
    mock.insert(
        new CustomersRow(
            new CustomersId(102L), "Charlie", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);

    var results =
        mock.select()
            .where(c -> c.customerId().greaterThan(new CustomersId(100L)))
            .orderBy(c -> c.name().asc())
            .toList(null);

    assertEquals(2, results.size());
    assertEquals("Bob", results.get(0).name());
    assertEquals("Charlie", results.get(1).name());
  }

  @Test
  public void testMockDSLCount() {
    var mock = createCustomersMock();

    mock.insert(
        new CustomersRow(
            new CustomersId(200L), "Count1", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        MockConnection.instance);
    mock.insert(
        new CustomersRow(
            new CustomersId(201L), "Count2", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        MockConnection.instance);
    mock.insert(
        new CustomersRow(
            new CustomersId(202L), "Other", Optional.empty(), LocalDateTime.of(2025, 1, 1, 0, 0)),
        MockConnection.instance);

    var count =
        mock.select()
            .where(c -> c.name().like("Count%", Bijection.asString()))
            .count(MockConnection.instance);

    assertEquals(2, count);
  }

  @Test
  public void testMockDSLLimit() {
    var mock = createCustomersMock();
    for (int i = 0; i < 10; i++) {
      mock.insert(
          new CustomersRow(
              new CustomersId(300L + i),
              "Limit" + i,
              Optional.empty(),
              LocalDateTime.of(2025, 1, 1, 0, 0)),
          null);
    }
    var results =
        mock.select()
            .where(c -> c.name().like("Limit%", Bijection.asString()))
            .limit(5)
            .toList(null);

    assertEquals(5, results.size());
  }

  @Test
  public void testMockIsolation() {
    var mock1 = createCustomersMock();
    var mock2 = createCustomersMock();

    mock1.insert(
        new CustomersRow(
            new CustomersId(400L),
            "Mock1 Only",
            Optional.empty(),
            LocalDateTime.of(2025, 1, 1, 0, 0)),
        null);

    assertTrue(mock1.selectById(new CustomersId(400L), null).isPresent());
    assertFalse(mock2.selectById(new CustomersId(400L), null).isPresent());
  }
}

File: testers/sqlite/kotlin/src/kotlin/testdb/MockRepoTest.kt
package testdb

import dev.typr.dslkt.MockConnection
import org.junit.Assert.*
import org.junit.Test
import testdb.customers.*
import testdb.departments.*
import testdb.userdefined.Email
import java.math.BigDecimal
import java.time.LocalDateTime

class MockRepoTest {

    private fun mkCustomersMock(): CustomersRepoMock = CustomersRepoMock({ unsaved ->
        CustomersRow(
            customerId = unsaved.customerId,
            name = unsaved.name,
            email = unsaved.email,
            createdAt = unsaved.createdAt.getOrElse { LocalDateTime.of(2025, 1, 1, 0, 0) }
        )
    })

    @Test
    fun testMockInsertAndSelect() {
        val mock = mkCustomersMock()
        val customer = CustomersRow(CustomersId(1L), "Mock User", Email("mock@test.com"), LocalDateTime.of(2025, 1, 1, 0, 0))
        val inserted = mock.insert(customer, MockConnection.instance)
        assertEquals("Mock User", inserted.name)
        val found = mock.selectById(CustomersId(1L), MockConnection.instance)
        assertNotNull(found)
        assertEquals("Mock User", found!!.name)
    }

    @Test
    fun testMockUpdate() {
        val mock = mkCustomersMock()
        val customer = CustomersRow(CustomersId(2L), "Original", null, LocalDateTime.of(2025, 1, 1, 0, 0))
        mock.insert(customer, MockConnection.instance)
        mock.update(customer.copy(name = "Updated"), MockConnection.instance)
        assertEquals("Updated", mock.selectById(CustomersId(2L), MockConnection.instance)!!.name)
    }

    @Test
    fun testMockDelete() {
        val mock = mkCustomersMock()
        mock.insert(CustomersRow(CustomersId(3L), "Del", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        assertNotNull(mock.selectById(CustomersId(3L), MockConnection.instance))
        mock.deleteById(CustomersId(3L), MockConnection.instance)
        assertNull(mock.selectById(CustomersId(3L), MockConnection.instance))
    }

    @Test
    fun testMockSelectAll() {
        val mock = mkCustomersMock()
        mock.insert(CustomersRow(CustomersId(10L), "U1", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(11L), "U2", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(12L), "U3", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        assertEquals(3, mock.selectAll(MockConnection.instance).size)
    }

    @Test
    fun testMockDSLLimit() {
        val mock = mkCustomersMock()
        for (i in 0 until 10) {
            mock.insert(CustomersRow(CustomersId(300L + i), "Limit$i", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        }
        val results = mock.select().where { it.name().like("Limit%") }.limit(5).toList(MockConnection.instance)
        assertEquals(5, results.size)
    }

    @Test
    fun testMockWithCompositeKey() {
        val mock = DepartmentsRepoMock()
        mock.insert(DepartmentsRow("MOCK_IT", "MOCK_US", "Mock IT US", BigDecimal("500000")), MockConnection.instance)
        val found = mock.selectById(DepartmentsId("MOCK_IT", "MOCK_US"), MockConnection.instance)
        assertNotNull(found)
        assertEquals("Mock IT US", found!!.deptName)
    }

    @Test
    fun testMockDSLQuery() {
        val mock = mkCustomersMock()
        mock.insert(CustomersRow(CustomersId(100L), "Alice", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(101L), "Bob", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(102L), "Charlie", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)

        val results = mock.select()
            .where { it.customerId().greaterThan(CustomersId(100L)) }
            .orderBy { it.name().asc() }
            .toList(MockConnection.instance)

        assertEquals(2, results.size)
        assertEquals("Bob", results[0].name)
        assertEquals("Charlie", results[1].name)
    }

    @Test
    fun testMockDSLCount() {
        val mock = mkCustomersMock()
        mock.insert(CustomersRow(CustomersId(200L), "Count1", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(201L), "Count2", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        mock.insert(CustomersRow(CustomersId(202L), "Other", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)

        val count = mock.select()
            .where { it.name().like("Count%") }
            .count(MockConnection.instance)
        assertEquals(2, count)
    }

    @Test
    fun testMockIsolation() {
        val mock1 = mkCustomersMock()
        val mock2 = mkCustomersMock()
        mock1.insert(CustomersRow(CustomersId(400L), "M1", null, LocalDateTime.of(2025, 1, 1, 0, 0)), MockConnection.instance)
        assertNotNull(mock1.selectById(CustomersId(400L), MockConnection.instance))
        assertNull(mock2.selectById(CustomersId(400L), MockConnection.instance))
    }
}
